Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,226 @@
+    errorPoly = fromInteger (1 `shiftL` (2 * dim)) :: F2Poly
+    dim = size matrix
+
+berlekampMassey :: Int -> F2Poly -> F2Poly -> F2Poly
+berlekampMassey dim = go 1 0
+  where
+    -- Is there a better way to implement recursion in this situation?
+    go :: F2Poly -> F2Poly -> F2Poly ->F2Poly -> F2Poly
+    go oneBefore twoBefore a b
+      | U.length (unF2Poly b) <= dim = oneBefore
+      -- Updated value is given by @twoBefore - oneBefore * q@
+      | otherwise                    = go (twoBefore - oneBefore * q) oneBefore b r
+        where
+          (q, r) = quotRem a b
+
+-- The input is a matrix B and a random vector z
+generateData :: KnownNat k => SBMatrix k -> DBVector k -> DBVector k -> F2Poly
+generateData matrix z x = toF2Poly $ U.fromList $ reverse $ map (x `dot`) matrixPowers
+  where
+    matrixPowers = L.take (2 * size matrix) $ L.iterate (matrix `mult`) z
+
+-- Infinite lists of random DBVectors.
+getRandomDBVectors :: KnownNat k => Int -> Double -> StdGen -> [DBVector k]
+getRandomDBVectors numberOfColumns density gen = go $ randomRs (0, 1) gen
+  where
+    go :: KnownNat k => [Double] -> [DBVector k]
+    go list = newVector `seq` newVector : go backOfList
+      where
+        newVector = DBVector $ fromJust $ SU.fromList $ map (\d -> Bit (d > density)) frontOfList
+        (frontOfList, backOfList) = L.splitAt numberOfColumns list
+
+-- Input number of columns of matrix and sparsity coefficient. It returns a random matrix.
+testLinearSolver :: Int -> Double -> Bool
+testLinearSolver dim density = case someNatVal (fromIntegral dim) of
+  SomeNat (_ :: Proxy dim) -> let sol :: DBVector dim = linearSolve mat in
+    mat `mult` sol == mempty
+      where
+        mat = SBMatrix $ fromJust $ SV.fromList listOfColumns
+        -- -2 is arbitrary. It means that the number of rows is at most one less than
+        -- the number of columns
+        listOfColumns = L.take dim $ getRandomSBVectors (dim - 2) density $ mkStdGen $ fromIntegral $ unsafePerformIO getMonotonicTimeNSec
+
+getRandomSBVectors :: KnownNat k => Int -> Double -> StdGen -> [SBVector k]
+getRandomSBVectors numberOfRows density gen = go randomEntries
+  where
+    randomEntries = map (< density) $ randomRs (0, 1) gen
+    go :: KnownNat k => [Bool] -> [SBVector k]
+    go list = newVector `seq` (newVector : go backOfList)
+      where
+        newVector = SBVector (U.fromList listOfEntries)
+        listOfEntries = map fst $ filter snd $ zip [minBound..maxBound] frontOfList
+        (frontOfList, backOfList) = L.splitAt numberOfRows list
+
+
+--------------------------------------------------------------------------------
+
+testGauss :: Int -> Double -> Bool
+testGauss dim density = sol /= S.empty
+  where
+    sol = gaussianElimination mat
+    mat = map (\(a, b) -> (S.singleton a, b)) $ zip [0..] listOfColumns
+    listOfColumns = L.take dim $ getRandomIntSets (dim - 2) density $ mkStdGen $ fromIntegral $ unsafePerformIO getMonotonicTimeNSec
```

Comment:
Unfortunately, `GHC.Clock` appeared only in `base-4.11` and so requires GHC >= 8.4. I suggest we use [`System.CPUTime.getCPUTime`](http://hackage.haskell.org/package/base-4.10.0.0/docs/System-CPUTime.html) as a source of entropy.